### PR TITLE
Resource support

### PR DESCRIPTION
**Pull request by runzexia**

### CODE REVIEW

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceManager.java [hunk 1]
```diff
@@ -6,23 +6,21 @@
...
+import java.io.PrintStream;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 
 import static com.google.common.base.Preconditions.checkNotNull;
 import static com.google.common.base.Preconditions.checkState;
 
 public abstract class ResourceManager {
-    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private PrintStream consoleLogger = System.out;
```
Comment (runzexia) [inline src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceManager.java]: change `logger` to  `consoleLogger` 
Print to jenkins console

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceManager.java [hunk 2]
```diff
@@ -118,21 +106,50 @@ final void createOrApply() throws IOException {
...
 
-    protected void handleApiException(ApiException e) {
+
+    /**
+     * Func to handle ApiException , print out the contents of the exception
+     * and throw a RuntimeException to abort the pipeline except NotFound Condition.
+     * @param e kubernetes ApiException
+     * @throws RuntimeException
+     */
+    protected void handleApiExceptionExceptNotFound(ApiException e) throws RuntimeException {
```
Comment (runzexia) [inline src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceManager.java]: 404 does not need to interrupt the pipeline when checking if the object exists

**Comment by runzexia** [comment]
I think I can now do some code review.
There are many codes in this PR, but most of them are similar.

Three of these resources are special: `V1Job`, `V1Pod`, `V1Statefulset`.
These resources cannot be directly put into the kubernetes apiserver.
Some default values are not allowed to be changed. 
So I get these defaults through server dryRun and construct putObject

**Comment by gavinfish** [comment]
@runzexia  sorry for my late response for this plugin since I have limit bandwidth for Jenkins related staff recently.
I think we are really close to release a new version to introduce the new SDK. I have run some simple tests with some kind of kubernetes resources on the latest branch and it works fine.

Still more review work and tests are needed before we release it.

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceManager.java (hunk 1 above)
Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceManager.java]: Why do we need this change? `Logger` works fine with Jenkins system.

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceManager.java [hunk 3]
```diff
@@ -84,23 +82,13 @@ final String getKind() {
...
-         * @throws IOException if we cannot find the resource in the cluster when we apply the configuration
          */
-        final void createOrApply() throws IOException {
+        final void createOrApply() {
             T original = getCurrentResource();
             T current = get();
             T updated;
             if (original != null) {
                 updated = applyResource(original, current);
-                if (updated == null) {
```
Comment (gavinfish) [inline src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceManager.java]: No more need nullPointer check?

Comment (runzexia) [reply src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceManager.java]: If an `ApiException` occurs, a `RuntimeException` is thrown to interrupt the pipeline, so no `null `value is returned.

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceManager.java (hunk 1 above)
Comment (runzexia) [reply src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceManager.java]: The original logger will print the log to the Jenkins system log, which prints the content to the Jenkins pipeline.

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java [hunk 4]
```diff
@@ -300,9 +374,17 @@ V1Service applyResource(V1Service original, V1Service current) {
             // this should be no-op, keep it in case current.getSpec().getPorts() behavior changes in future
             current.getSpec().setPorts(currentPorts);
 
+            // update svc need to set resourceVersion
+            current.getMetadata().setResourceVersion(original.getMetadata().getResourceVersion());
+
+            // clusterIP is immutable, if the clusterIP is empty, it is set to the original clusterIP
+            if (!StringUtils.isNotBlank(current.getSpec().getClusterIP())) {
```
[the remark this answers is not in the document]
Comment (runzexia) [reply src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java]: fix

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java [hunk 5]
```diff
@@ -6,40 +6,116 @@
...
+            checkNotNull(yaml.getInitContainers());
+            checkNotNull(target.getInitContainers());
+            // should not change size
+            if (origin.getInitContainers().size() != yaml.getInitContainers().size()
+                    || origin.getInitContainers().size() != target.getInitContainers().size()) {
+                return;
+            }
+            for (int i = 0; i < origin.getInitContainers().size(); i++) {
+                if (yaml.getInitContainers().get(i) == null) {
+                    target.getInitContainers().get(i).imagePullPolicy(
```
[the remark this answers is not in the document]
Comment (runzexia) [reply src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java]: Yes The container will be sorted.
I should check if imagePullPolicy is empty. If it is not empty, the user has modified it.

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java [hunk 6]
```diff
@@ -109,7 +109,7 @@ private static void recoverPodImagePullPolicy(V1PodSpec origin, V1PodSpec yaml,
                 return;
             }
             for (int i = 0; i < origin.getContainers().size(); i++) {
-                if (yaml.getContainers().get(i) == null) {
+                if (StringUtils.isBlank(yaml.getInitContainers().get(i).getImagePullPolicy())) {
```
Comment (gavinfish) [inline src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java]: Typo? not init containers here

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java [hunk 7]
```diff
@@ -337,29 +419,83 @@ void notifyUpdate(V1Service original, V1Service current) {
...
+                    name(getName()).
+                    creationTimestamp(original.getMetadata().getCreationTimestamp()).
+                    selfLink(original.getMetadata().getSelfLink()).
+                    uid(original.getMetadata().getUid()).
+                    resourceVersion(original.getMetadata().getUid()).
+                    ownerReferences(original.getMetadata().getOwnerReferences()).
+                    generateName(null);
+            // Recover label/selector
+            String controllerUid = original.getMetadata().getLabels().get(KUBERNETES_CONTROLLER_UID_FIELD);
+            dryRunRes.getMetadata().getLabels().put(KUBERNETES_CONTROLLER_UID_FIELD, controllerUid);
```
Comment (gavinfish) [inline src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java]: Any resources to support these implementation? How can we know which properties should we reset here? I am afraid we may miss something here.

Comment (runzexia) [reply src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java]: Have added comments in code

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java [hunk 8]
```diff
@@ -337,29 +420,89 @@ void notifyUpdate(V1Service original, V1Service current) {
...
+         * This method is a workaround that populates the default value with the server dryRun implementation.
+         * https://kubernetes.io/blog/2019/01/14/apiserver-dry-run-and-kubectl-diff/
+         *
+         * @param original Current Object in Cluster
+         * @return Job With Default Value
+         */
+        V1Job getPutObject(V1Job original) {
+            // Clone Object to avoid modifications to the original object.
+            V1Job dryRunReq = KubernetesJsonUtils.getKubernetesJson().deserialize(
+                    KubernetesJsonUtils.getKubernetesJson().serialize(get()), V1Job.class);
```
Comment (gavinfish) [inline src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java]: Change method to `getPutObject(V1Job original, V1Job current)` and replace `get()` with current passed by `applyResource`?

**Pull request merged**